diff --git a/releasetools.py b/releasetools.py
index 8b5b1e2..cde92b1 100644
--- a/releasetools.py
+++ b/releasetools.py
@@ -13,21 +13,29 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.

+import hashlib
 import common
+import re

 def FullOTA_InstallEnd(info):
-  OTA_InstallEnd(info)
+  input_zip = info.input_zip
+  OTA_InstallEnd(info, input_zip)
   return

 def IncrementalOTA_InstallEnd(info):
-  OTA_InstallEnd(info)
+  input_zip = info.target_zip
+  OTA_InstallEnd(info, input_zip)
   return

-def AddImage(info, basename, dest):
+def AddImage(info, input_zip, basename, dest):
   name = basename
-  data = info.input_zip.read("IMAGES/" + basename)
-  common.ZipWriteStr(info.output_zip, name, data)
-  info.script.AppendExtra('package_extract_file("%s", "%s");' % (name, dest))
+  path = "IMAGES/" + name
+  if path not in input_zip.namelist():
+    return
+
+  data = input_zip.read(path)
+  common.ZipWriteStr(info.output_zip, basename, data)
+  info.script.AppendExtra('package_extract_file("%s", "%s");' % (basename, dest))

 def OTA_InstallEnd(info):
   info.script.Print("Patching firmware images...")
