diff --git a/build/tasks/kernel.mk b/build/tasks/kernel.mk
index c732a07..f74abe7 100644
--- a/build/tasks/kernel.mk
+++ b/build/tasks/kernel.mk
@@ -112,13 +112,6 @@ DTBS_OUT := $(DTB_OUT)/out
 endif
 KERNEL_CONFIG := $(KERNEL_OUT)/.config
 KERNEL_RELEASE := $(KERNEL_OUT)/include/config/kernel.release
-ifeq ($(BOARD_USES_QCOM_HARDWARE),true)
-ifeq ($(call is-version-greater-or-equal,$(TARGET_KERNEL_VERSION),5.15),true)
-ifeq ($(call is-version-lower-or-equal,$(TARGET_KERNEL_VERSION),6.1),true)
-GKI_SUFFIX := /$(shell echo android$(PLATFORM_VERSION)-$(TARGET_KERNEL_VERSION))
-endif
-endif
-endif

 ifeq ($(KERNEL_ARCH),x86_64)
 KERNEL_DEFCONFIG_ARCH := x86

