diff --git a/core/java/android/view/animation/AnimationUtils.java b/core/java/android/view/animation/AnimationUtils.java
index 8ecd57179b..a393960877 100644
--- a/core/java/android/view/animation/AnimationUtils.java
+++ b/core/java/android/view/animation/AnimationUtils.java
@@ -16,6 +16,10 @@
 
 package android.view.animation;
 
+import android.os.SystemProperties;
+import android.util.DisplayMetrics;
+import com.android.internal.R;
+
 import static android.view.flags.Flags.FLAG_EXPECTED_PRESENTATION_TIME_READ_ONLY;
 import static android.view.flags.Flags.expectedPresentationTimeReadOnly;
 
@@ -226,6 +230,20 @@ public class AnimationUtils {
     public static Animation loadAnimation(Context context, @AnimRes int id)
             throws NotFoundException {
 
+        if (SystemProperties.getBoolean("persist.sys.activity_anim_perf_override", false)) {
+            ActivityAnimations.maybeInit(context);
+            switch (id) {
+                case R.anim.activity_open_enter:
+                    return ActivityAnimations.getOpenEnter();
+                case R.anim.activity_open_exit:
+                    return ActivityAnimations.getOpenExit();
+                case R.anim.activity_close_enter:
+                    return ActivityAnimations.getCloseEnter();
+                case R.anim.activity_close_exit:
+                    return ActivityAnimations.getCloseExit();
+            }
+        }
+
         XmlResourceParser parser = null;
         try {
             parser = context.getResources().getAnimation(id);
@@ -506,4 +524,100 @@ public class AnimationUtils {
         }
         return interpolator;
     }
+
+    /** @hide */
+    public final class ActivityAnimations {
+
+        private static Animation sOpenEnter;
+        private static Animation sOpenExit;
+        private static Animation sCloseEnter;
+        private static Animation sCloseExit;
+
+        private static Interpolator sFastOutExtraSlowInInterpolator;
+
+        private static final float DISTANCE = 0.1f;
+
+        private ActivityAnimations() {}
+
+        /** @hide */
+        public static void maybeInit(Context context) {
+            if (sFastOutExtraSlowInInterpolator == null) {
+                sFastOutExtraSlowInInterpolator = AnimationUtils.loadInterpolator(
+                        context, R.interpolator.fast_out_extra_slow_in);
+            }
+        }
+
+        private static class ActivityAnimFactory {
+            private float fromX = 0f, toX = 0f;
+            private long duration = 200L;
+
+            public ActivityAnimFactory fromX(float ratio) {
+                this.fromX = ratio;
+                return this;
+            }
+
+            public ActivityAnimFactory toX(float ratio) {
+                this.toX = ratio;
+                return this;
+            }
+
+            public Animation build() {
+                AnimationSet animationSet = new AnimationSet(false);
+                TranslateAnimation slide = new TranslateAnimation(
+                        Animation.RELATIVE_TO_SELF, fromX,
+                        Animation.RELATIVE_TO_SELF, toX,
+                        Animation.RELATIVE_TO_SELF, 0f,
+                        Animation.RELATIVE_TO_SELF, 0f
+                );
+                slide.setDuration(duration);
+                slide.setInterpolator(sFastOutExtraSlowInInterpolator);
+                animationSet.addAnimation(slide);
+                return animationSet;
+            }
+        }
+
+        /** @hide */
+        public static Animation getOpenEnter() {
+            if (sOpenEnter == null) {
+                sOpenEnter = new ActivityAnimFactory()
+                        .fromX(1.0f)
+                        .toX(0.0f)
+                        .build();
+            }
+            return sOpenEnter;
+        }
+
+        /** @hide */
+        public static Animation getOpenExit() {
+            if (sOpenExit == null) {
+                sOpenExit = new ActivityAnimFactory()
+                        .fromX(0.0f)
+                        .toX(-DISTANCE)
+                        .build();
+            }
+            return sOpenExit;
+        }
+
+        /** @hide */
+        public static Animation getCloseEnter() {
+            if (sCloseEnter == null) {
+                sCloseEnter = new ActivityAnimFactory()
+                        .fromX(-DISTANCE)
+                        .toX(0.0f)
+                        .build();
+            }
+            return sCloseEnter;
+        }
+
+        /** @hide */
+        public static Animation getCloseExit() {
+            if (sCloseExit == null) {
+                sCloseExit = new ActivityAnimFactory()
+                        .fromX(0.0f)
+                        .toX(1.0f)
+                        .build();
+            }
+            return sCloseExit;
+        }
+    }
 }
